File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n=nums.size();
        int start=0,end=n-1,mid;
        while(start<=end){
            mid=(end+start)/2;
            if(nums[mid]==target) return mid;
            if(nums[mid]>=nums[start]){
                if(target<=nums[mid] && target>=nums[start]){
                    end=mid-1;
                }
                else{
                    start=mid+1;
                }
            }
            else{
                if(target<=nums[end] && target>=nums[mid]){
                    start=mid+1;
                }
                else{
                    end=mid-1;
                }
            }
        }
        return -1;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        l,r = 0,n-1
        while l<=r:
            m = (l+r)//2
            if target == nums[m]:
                return m
            if nums[m]>=nums[l]:
                if target<nums[m] and target>=nums[l]:
                    r = m-1
                else:
                    l = m+1
            else:
                if target>nums[m] and target<=nums[r]:
                    l=m+1
                else:
                    r=m-1
        return -1
